src/dnf-plugins/product-id/product-id.c

```diff
@@ -216,6 +219,80 @@ void requestProductIdMetadata(DnfContext *dnfContext) {
     }
 }
 
+/**
+ * This function tries to get list of installed product certificates from installed product certificates.
+ *
+ * @param repos Array of DnfRepos
+ * @param enabledRepoProductId Array of RepoProductId. It should be empty, when this function is called
+ * @param productDb Pointer at ProductDb
+ * @return
+ */
+int getInstalledProductCerts(gchar *certDir, GPtrArray *repos, GPtrArray *enabledRepoProductId, ProductDb *productDb) {
+    int ret = 0;
+    // "Open" directory with product certificates
+    GError *tmp_err = NULL;
+    GDir* productDir = g_dir_open(certDir, 0, &tmp_err);
+    if (productDir != NULL) {
+        const gchar *file_name = NULL;
+        do {
+            // Read all files in the directory. When file_name is NULL, then
+            // it usually means that there is no more file.
+            file_name = g_dir_read_name(productDir);
+            if(file_name != NULL) {
+                if(g_str_has_suffix(file_name, ".pem") == TRUE) {
+                    gchar *product_id = g_strndup(file_name, strlen(file_name) - 4);
+                    debug("Productid cert: %s", file_name);
+                    gboolean is_num = TRUE;
+                    // Test if string represents number
+                    for(size_t i=0; i<strlen(product_id); i++) {
+                        if (g_ascii_isdigit(product_id[i]) != TRUE) {
+                            is_num = FALSE;
+                            break;
+                        }
+                    }
+                    if (is_num != TRUE) {
+                        debug("Name of product certificate is wrong (not digits only): %s. Skipping.", file_name);
+                    }
+                        // When product certificate is not in the hash table of active repositories
+                        // then it is IMHO possible to remove this product certificate
+                    else if (hasProductId(productDb, product_id)) {
+                        gchar *abs_file_name = g_strconcat(certDir, file_name, NULL);
+                        GSList *repoIds = getRepoIds(productDb, product_id);
+                        if (repoIds != NULL) {
+                            debug("Existing product ID: %s has following repos:", product_id);
+                            GSList *iterator = NULL;
+                            // The list usually contains only one record
+                            for (iterator = repoIds; iterator; iterator = iterator->next) {
+                                debug("\tRepo: %s", iterator->data);
+                                for (guint i = 0; i < repos->len; i++) {
+                                    DnfRepo *repo = g_ptr_array_index(repos, i);
+                                    if (g_strcmp0(dnf_repo_get_id(repo), iterator->data) == 0) {
```

Comment:
I realize this is merged already, but at some point I think it would be good to go back and break this method up a little more.  If my count is correct, this line is 7 levels deep which is a little bit too much for my brain to keep track of.

Reply:
Good point. Thanks. I just do some refactoring of the code. I will add it to next PR.